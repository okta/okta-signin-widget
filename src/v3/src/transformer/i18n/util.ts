/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

import {
  getI18nKey,
  getI18NParams,
} from '../../../../v2/ion/i18nUtils';
import {
  FieldElement, TokenReplacement, UISchemaElement,
} from '../../types';
import { loc } from '../../util';

export const addTranslation = ({
  element,
  name,
  i18nKey,
  params,
  defaultValue = '',
  tokenReplacement,
  noTranslate,
}: {
  element: UISchemaElement;
  name: string;
  i18nKey: string;
  params?: any;
  defaultValue?: string;
  tokenReplacement?: TokenReplacement,
  noTranslate?: boolean,
}): void => {
  // TODO: change translations to required field with default value
  // eslint-disable-next-line no-param-reassign
  element.translations = element.translations || [];
  const useDefault = !i18nKey || (noTranslate && !!defaultValue);
  element.translations.push({
    name,
    i18nKey,
    value: useDefault ? defaultValue : loc(i18nKey, 'login', params, tokenReplacement),
    noTranslate,
  });
};

export const addLabelTranslationToFieldElement = (
  transaction: IdxTransaction,
  element: FieldElement,
): void => {
  const { nextStep } = transaction;
  const { relatesTo, name: stepName } = nextStep!;
  const authenticatorKey = relatesTo?.value?.key;
  const fieldName = element.options.inputMeta.name;
  const path = authenticatorKey
    ? `${stepName}.${authenticatorKey}.${fieldName}`
    : `${stepName}.${fieldName}`;
  const i18nKey = getI18nKey(path);
  const params = getI18NParams(nextStep, authenticatorKey);
  if (i18nKey || element.label) {
    const noTranslate = element.options.inputMeta.customLabel;
    addTranslation({
      element, name: 'label', i18nKey, params, defaultValue: element.label, noTranslate,
    });
  }
};

export const addOptionalLabelTranslationToFieldElement = (
  element: FieldElement,
): void => {
  if (element.options.inputMeta.required === false) {
    addTranslation({
      element, name: 'optionalLabel', i18nKey: 'oie.form.field.optional',
    });
  }
};
