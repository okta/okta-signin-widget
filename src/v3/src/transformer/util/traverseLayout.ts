/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  FormBag,
  StepperLayout,
  UISchemaElement,
  UISchemaLayout,
  UISchemaLayoutType,
} from '../../types';

type PredicateFn = (uischema: UISchemaElement) => boolean;
type CallbackFn = (uischema: UISchemaElement) => void;

type Options = {
  layout: FormBag['uischema'];
  predicate: PredicateFn;
  callback: CallbackFn;
};

export const traverseLayout = (options: Options) => {
  const fn = (
    layout: FormBag['uischema'],
    predicateFn: PredicateFn,
    callback: CallbackFn,
  ) => {
    layout.elements.forEach((element) => {
      if (predicateFn(element)) {
        callback(element);
      }

      const { type } = element;

      if (type === UISchemaLayoutType.STEPPER) {
        (element as StepperLayout).elements
          .forEach((el) => fn(el, predicateFn, callback));
        return;
      }

      if ([UISchemaLayoutType.HORIZONTAL, UISchemaLayoutType.VERTICAL]
        .includes(type as UISchemaLayoutType)) {
        fn(element as UISchemaLayout, predicateFn, callback);
      }
    });
  };

  fn(options.layout, options.predicate, options.callback);
};
