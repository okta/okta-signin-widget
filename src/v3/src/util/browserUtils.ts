/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export const isAndroid = (): boolean => (
  // Windows Phone also contains "Android"
  /android/i.test(navigator.userAgent)
    && !/windows phone/i.test(navigator.userAgent)
);

export const isIOS = (): boolean => (
  // iOS detection from: http://stackoverflow.com/a/9039885/177710
  /iPad|iPhone|iPod/i.test(navigator.userAgent)
    // iPad on iOS 13 detection
    || (navigator.userAgent.includes('Mac') && 'ontouchend' in document)
);

export const isAndroidOrIOS = (): boolean => isAndroid() || isIOS();
