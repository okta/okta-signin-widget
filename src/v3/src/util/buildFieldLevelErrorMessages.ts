/*
 * Copyright (c) 2024-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { WidgetMessage } from '../types';

type FieldLevelErrorMessages = {
  errorMessage?: string;
  errorMessageList?: string[];
};

export const buildFieldLevelErrorMessages = (
  errors: WidgetMessage[] | undefined,
) : FieldLevelErrorMessages => {
  const hasErrors = typeof errors !== 'undefined' && errors.length > 0;

  if (hasErrors) {
    if (errors.length === 1) {
      const error = errors[0];
      // Covers nested messages, e.g. password requirements not met
      // The SIW does not have messages nested more than once
      if (Array.isArray(error.message)) {
        return {
          errorMessage: error.description,
          errorMessageList: error.message
            .filter((msg) => typeof msg.message === 'string')
            .map((msg) => msg.message as string),
        };
      }
      return { errorMessage: error.message };
    }
    return {
      errorMessageList: errors
        .filter((error) => typeof error.message === 'string')
        .map((error) => error.message as string),
    };
  }

  return {};
};
