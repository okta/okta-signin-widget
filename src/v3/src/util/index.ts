/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export * from './browserUtils';
export * from './buildErrorMessageIds';
export * from './clipboard';
export * from './cookieUtils';
export * from './environmentUtils';
export * from './flattenInputs';
export * from './formatError';
export * from './formUtils';
export * from './generateRandomString';
export * from './getAuthenticatorKey';
export * from './getCurrentAuthenticator';
export * from './getElementKey';
export * from './getImmutableData';
export * from './getTranslation';
export * from './getValidationMessages';
export * from './idxUtils';
export * from './isInteractiveElement';
export * from './isPasswordRecovery';
export * from './isPollingStep';
export * from './languageUtils';
export * from './locUtil';
export * from './passwordUtils';
export * from './removeFieldLevelMessages';
export * from './resetMessagesToInputs';
export * from './settingsUtils';
export * from './shouldShowCancelLink';
export * from './toNestedObject';
export * from './webauthnUtils';
