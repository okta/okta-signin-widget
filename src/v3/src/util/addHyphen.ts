/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
*/
import { KeyboardEvent } from "src/types"

export const addHyphen = (e: KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>): void => {
    const currentVal = e.currentTarget.value;
    // Add hyphen after 4th character. includes() check allows deletion of hyphen and prevents double hyphens
    if (currentVal && currentVal.length === 4 && !['Backspace', 'Delete', '-'].includes(e.key)) {
      e.currentTarget.value = currentVal.concat('-');
    } else if (currentVal && currentVal.length > 4 && currentVal.charAt(4) !== '-') {
      // If user types too fast, the above concatenation can be skipped. This retroactively adds a hyphen if it wasn't added
      e.currentTarget.value = currentVal.slice(0, 4) + '-' + currentVal.slice(4);
    }
}
