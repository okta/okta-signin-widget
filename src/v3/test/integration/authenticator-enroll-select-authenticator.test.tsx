/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { createAuthJsPayloadArgs, setup } from './util';

import mockResponse from '../../src/mocks/response/idp/idx/authenticator-enroll-select-authenticator.json';

describe('authenticator-enroll-select-authenticator', () => {
  it('renders form', async () => {
    const { container, findByText } = await setup({ mockResponse });
    await findByText(/Set up security methods/);
    await findByText(/Set up required/);
    expect(container).toMatchSnapshot();
  });

  describe('send correct payload', () => {
    it('when select google authenticator', async () => {
      const {
        authClient,
        user,
        findByTestId,
      } = await setup({ mockResponse });

      const authenticatorButton = await findByTestId('google_otp');
      await user.click(authenticatorButton);
      expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
        ...createAuthJsPayloadArgs('POST', 'idp/idx/credential/enroll', {
          authenticator: { id: 'aut11ceMaP0B0EzMI0g4' },
        }),
      );
    });

    it('should only send api request once when double clicked', async () => {
      const {
        authClient,
        user,
        findByTestId,
      } = await setup({ mockResponse });

      const authenticatorButton = await findByTestId('google_otp');
      await user.dblClick(authenticatorButton);
      expect(authClient.options.httpRequestClient).toHaveBeenCalledTimes(1);
    });
  });

  // TODO: Additional tests and flows can be added when testing monolith scenarios
  /*
    * Scenarios where the stateToken is provided as a prop to the Widget
    * In these cases, the stateHandle should be passed to auth-js and should not throw an exception
    * If this exception is thrown "Unable to proceed: saved transaction could not be loaded"
    * it is because either SIW isnt providing the stateHandle or auth-js somehow lost the
    * provided stateHandled in translation.
    */
  describe('When StateToken is provided to widget', () => {
    it('should send correct payload when clicking cancel button', async () => {
      const {
        authClient,
        user,
        findByTestId,
      } = await setup(
        {
          mockResponse,
          widgetOptions: {
            stateToken: '1234567890abcdefghij',
            authScheme: 'false',
          },
        },
      );

      const cancelBtn = await findByTestId('cancel');
      await user.click(cancelBtn);

      expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
        ...createAuthJsPayloadArgs('POST', 'idp/idx/cancel'),
      );
    });
  });

  describe('useInteractionCodeFlow', () => {
    it('should restart transaction when clicking "Back to sign in" link', async () => {
      const {
        authClient,
        user,
        findByTestId,
      } = await setup(
        {
          mockResponse,
          widgetOptions: {
            authScheme: 'oauth2',
          },
        },
      );

      const cancelBtn = await findByTestId('cancel');
      await user.click(cancelBtn);
      expect(authClient.options.httpRequestClient).toHaveBeenNthCalledWith(1,
        'POST', 'http://localhost:3000/oauth2/default/v1/interact', expect.any(Object));
      expect(authClient.options.httpRequestClient).toHaveBeenNthCalledWith(2,
        ...createAuthJsPayloadArgs(
          'POST', 'idp/idx/introspect', {
            interactionHandle: 'fake-interactionhandle',
          },
          'application/ion+json; okta-version=1.0.0',
          'application/ion+json; okta-version=1.0.0',
        ));
    });
  });
});
