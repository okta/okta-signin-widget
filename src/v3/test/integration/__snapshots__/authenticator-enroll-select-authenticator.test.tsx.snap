// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticator-enroll-select-authenticator renders form 1`] = `
<div>
  <div
    class="MuiScopedCssBaseline-root emotion-0"
  >
    <span>
      <main
        class="auth-container main-container mainViewContainer MuiBox-root emotion-1"
        data-commit="b9bbc0140703c3fbf0e2e58920362e70"
        data-version="0.0.0"
        id="okta-sign-in"
      >
        <div
          class="siwContainer MuiBox-root emotion-2"
        >
          <div
            class="okta-sign-in-header auth-header siwHeader"
          >
            <h1
              class="MuiTypography-root MuiTypography-h1 emotion-3"
            />
          </div>
          <div
            data-test="sentinelStart"
            tabindex="0"
          />
          <div
            class="MuiBox-root emotion-4"
            tabindex="-1"
          >
            <div
              class="identifier-container MuiBox-root emotion-5"
            >
              <div
                class="identifierContainer MuiBox-root emotion-6"
              >
                <span
                  class="userIconContainer MuiBox-root emotion-7"
                >
                  <svg
                    class="ods-2icygl"
                    fill="none"
                    role="presentation"
                    viewBox="0 0 16 16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M10 3V5C10 6.10457 9.10457 7 8 7C6.89543 7 6 6.10457 6 5V3C6 1.89543 6.89543 1 8 1C9.10457 1 10 1.89543 10 3ZM5 3C5 1.34315 6.34315 0 8 0C9.65685 0 11 1.34315 11 3V5C11 6.65685 9.65685 8 8 8C6.34315 8 5 6.65685 5 5V3ZM4.1305 10.0742C4.26803 10.0181 4.459 10 5.41376 10H10.5862C11.541 10 11.732 10.0181 11.8695 10.0742C12.0299 10.1398 12.1706 10.2459 12.2777 10.3821C12.3695 10.4989 12.4393 10.6776 12.7016 11.5956L13.6743 15H2.32573L3.29841 11.5956C3.5607 10.6776 3.63054 10.4989 3.72233 10.3821C3.82941 10.2459 3.97006 10.1398 4.1305 10.0742ZM13.6631 11.3209L14.7143 15L15 16H13.96H2.04002H1L1.28571 15L2.33689 11.3209C2.57458 10.489 2.69343 10.073 2.93606 9.76423C3.15022 9.49171 3.43152 9.27953 3.75239 9.14848C4.11592 9 4.54854 9 5.41376 9H10.5862C11.4515 9 11.8841 9 12.2476 9.14848C12.5685 9.27953 12.8498 9.49171 13.0639 9.76423C13.3066 10.073 13.4254 10.489 13.6631 11.3209Z"
                      fill="currentColor"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
                <span
                  class="identifier identifierSpan MuiBox-root emotion-7"
                  data-se="identifier"
                >
                  testUser@okta.com
                </span>
              </div>
            </div>
            <form
              class="o-form"
              data-se="form"
              novalidate=""
            >
              <div
                class="MuiBox-root emotion-9"
              >
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="MuiBox-root emotion-11"
                  >
                    <h2
                      class="MuiTypography-root MuiTypography-h3 emotion-12"
                      data-se="o-form-head"
                    >
                      Set up security methods
                    </h2>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="MuiBox-root emotion-11"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph emotion-15"
                      data-se="o-form-explain"
                    >
                      Security methods help protect your account by ensuring only you have access.
                    </p>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="MuiBox-root emotion-11"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph emotion-15"
                      data-se="o-form-explain"
                    >
                      Set up required
                    </p>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="authButton MuiBox-root emotion-20"
                    data-se="authenticator-button"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiBox-root emotion-7"
                      data-se="authenticator-icon"
                    >
                      <div
                        class="iconContainer mfa-okta-password MuiBox-root emotion-7"
                      >
                        <mocksvgicon />
                      </div>
                    </div>
                    <div
                      class="infoSection MuiBox-root emotion-7"
                    >
                      <div
                        class="title MuiBox-root emotion-7"
                        data-se="authenticator-button-label"
                      >
                        Password
                      </div>
                      <div
                        class="description MuiBox-root emotion-7"
                        data-se="authenticator-button-description"
                      >
                        Choose a password for your account
                      </div>
                      <div
                        class="cta-button actionName MuiBox-root emotion-7"
                        data-se="okta_password"
                      >
                        <span
                          class="MuiBox-root emotion-7"
                          data-se="cta-button-label"
                        >
                          Set up
                        </span>
                        <mocksvgicon />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="authButton MuiBox-root emotion-20"
                    data-se="authenticator-button"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiBox-root emotion-7"
                      data-se="authenticator-icon"
                    >
                      <div
                        class="iconContainer mfa-okta-phone MuiBox-root emotion-7"
                      >
                        <mocksvgicon />
                      </div>
                    </div>
                    <div
                      class="infoSection MuiBox-root emotion-7"
                    >
                      <div
                        class="title MuiBox-root emotion-7"
                        data-se="authenticator-button-label"
                      >
                        Phone
                      </div>
                      <div
                        class="description MuiBox-root emotion-7"
                        data-se="authenticator-button-description"
                      >
                        Verify with a code sent to your phone
                      </div>
                      <div
                        class="cta-button actionName MuiBox-root emotion-7"
                        data-se="phone_number"
                      >
                        <span
                          class="MuiBox-root emotion-7"
                          data-se="cta-button-label"
                        >
                          Set up
                        </span>
                        <mocksvgicon />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="authButton MuiBox-root emotion-20"
                    data-se="authenticator-button"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiBox-root emotion-7"
                      data-se="authenticator-icon"
                    >
                      <div
                        class="iconContainer mfa-webauthn MuiBox-root emotion-7"
                      >
                        <mocksvgicon />
                      </div>
                    </div>
                    <div
                      class="infoSection MuiBox-root emotion-7"
                    >
                      <div
                        class="title MuiBox-root emotion-7"
                        data-se="authenticator-button-label"
                      >
                        Security Key or Biometric Authenticator
                      </div>
                      <div
                        class="description MuiBox-root emotion-7"
                        data-se="authenticator-button-description"
                      >
                        Use a security key or a biometric authenticator to sign in
                      </div>
                      <div
                        class="cta-button actionName MuiBox-root emotion-7"
                        data-se="webauthn"
                      >
                        <span
                          class="MuiBox-root emotion-7"
                          data-se="cta-button-label"
                        >
                          Set up
                        </span>
                        <mocksvgicon />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="authButton MuiBox-root emotion-20"
                    data-se="authenticator-button"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiBox-root emotion-7"
                      data-se="authenticator-icon"
                    >
                      <div
                        class="iconContainer mfa-okta-security-question MuiBox-root emotion-7"
                      >
                        <mocksvgicon />
                      </div>
                    </div>
                    <div
                      class="infoSection MuiBox-root emotion-7"
                    >
                      <div
                        class="title MuiBox-root emotion-7"
                        data-se="authenticator-button-label"
                      >
                        Security Question
                      </div>
                      <div
                        class="description MuiBox-root emotion-7"
                        data-se="authenticator-button-description"
                      >
                        Choose a security question and answer that will be used for signing in
                      </div>
                      <div
                        class="cta-button actionName MuiBox-root emotion-7"
                        data-se="security_question"
                      >
                        <span
                          class="MuiBox-root emotion-7"
                          data-se="cta-button-label"
                        >
                          Set up
                        </span>
                        <mocksvgicon />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="authButton MuiBox-root emotion-20"
                    data-se="authenticator-button"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiBox-root emotion-7"
                      data-se="authenticator-icon"
                    >
                      <div
                        class="iconContainer mfa-okta-verify MuiBox-root emotion-7"
                      >
                        <mocksvgicon />
                      </div>
                    </div>
                    <div
                      class="infoSection MuiBox-root emotion-7"
                    >
                      <div
                        class="title MuiBox-root emotion-7"
                        data-se="authenticator-button-label"
                      >
                        Okta Verify
                      </div>
                      <div
                        class="description MuiBox-root emotion-7"
                        data-se="authenticator-button-description"
                      >
                        Okta Verify is an authenticator app, installed on your phone, used to prove your identity
                      </div>
                      <div
                        class="cta-button actionName MuiBox-root emotion-7"
                        data-se="okta_verify"
                      >
                        <span
                          class="MuiBox-root emotion-7"
                          data-se="cta-button-label"
                        >
                          Set up
                        </span>
                        <mocksvgicon />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="authButton MuiBox-root emotion-20"
                    data-se="authenticator-button"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiBox-root emotion-7"
                      data-se="authenticator-icon"
                    >
                      <div
                        class="iconContainer mfa-google-auth MuiBox-root emotion-7"
                      >
                        <mocksvgicon />
                      </div>
                    </div>
                    <div
                      class="infoSection MuiBox-root emotion-7"
                    >
                      <div
                        class="title MuiBox-root emotion-7"
                        data-se="authenticator-button-label"
                      >
                        Google Authenticator
                      </div>
                      <div
                        class="description MuiBox-root emotion-7"
                        data-se="authenticator-button-description"
                      >
                        Enter a temporary code generated from the Google Authenticator app.
                      </div>
                      <div
                        class="cta-button actionName MuiBox-root emotion-7"
                        data-se="google_otp"
                      >
                        <span
                          class="MuiBox-root emotion-7"
                          data-se="cta-button-label"
                        >
                          Set up
                        </span>
                        <mocksvgicon />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="authButton MuiBox-root emotion-20"
                    data-se="authenticator-button"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiBox-root emotion-7"
                      data-se="authenticator-icon"
                    >
                      <div
                        class="iconContainer mfa-onprem MuiBox-root emotion-7"
                      >
                        <mocksvgicon />
                      </div>
                    </div>
                    <div
                      class="infoSection MuiBox-root emotion-7"
                    >
                      <div
                        class="title MuiBox-root emotion-7"
                        data-se="authenticator-button-label"
                      >
                        Atko Custom On-prem
                      </div>
                      <div
                        class="description MuiBox-root emotion-7"
                        data-se="authenticator-button-description"
                      >
                        Verify by entering a code generated by Atko Custom On-prem.
                      </div>
                      <div
                        class="cta-button actionName MuiBox-root emotion-7"
                        data-se="onprem_mfa-otp"
                      >
                        <span
                          class="MuiBox-root emotion-7"
                          data-se="cta-button-label"
                        >
                          Set up
                        </span>
                        <mocksvgicon />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="authButton MuiBox-root emotion-20"
                    data-se="authenticator-button"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiBox-root emotion-7"
                      data-se="authenticator-icon"
                    >
                      <div
                        class="iconContainer mfa-rsa MuiBox-root emotion-7"
                      >
                        <mocksvgicon />
                      </div>
                    </div>
                    <div
                      class="infoSection MuiBox-root emotion-7"
                    >
                      <div
                        class="title MuiBox-root emotion-7"
                        data-se="authenticator-button-label"
                      >
                        RSA SecurID
                      </div>
                      <div
                        class="description MuiBox-root emotion-7"
                        data-se="authenticator-button-description"
                      >
                        Verify by entering a code generated by RSA SecurID
                      </div>
                      <div
                        class="cta-button actionName MuiBox-root emotion-7"
                        data-se="rsa_token-otp"
                      >
                        <span
                          class="MuiBox-root emotion-7"
                          data-se="cta-button-label"
                        >
                          Set up
                        </span>
                        <mocksvgicon />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="authButton MuiBox-root emotion-20"
                    data-se="authenticator-button"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiBox-root emotion-7"
                      data-se="authenticator-icon"
                    >
                      <div
                        class="iconContainer mfa-duo MuiBox-root emotion-7"
                      >
                        <mocksvgicon />
                      </div>
                    </div>
                    <div
                      class="infoSection MuiBox-root emotion-7"
                    >
                      <div
                        class="title MuiBox-root emotion-7"
                        data-se="authenticator-button-label"
                      >
                        Duo Security
                      </div>
                      <div
                        class="description MuiBox-root emotion-7"
                        data-se="authenticator-button-description"
                      >
                        Verify your identity using Duo Security.
                      </div>
                      <div
                        class="cta-button actionName MuiBox-root emotion-7"
                        data-se="duo"
                      >
                        <span
                          class="MuiBox-root emotion-7"
                          data-se="cta-button-label"
                        >
                          Set up
                        </span>
                        <mocksvgicon />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="authButton MuiBox-root emotion-20"
                    data-se="authenticator-button"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiBox-root emotion-7"
                      data-se="authenticator-icon"
                    >
                      <div
                        class="iconContainer mfa-custom-factor MuiBox-root emotion-7"
                      >
                        <mocksvgicon />
                      </div>
                    </div>
                    <div
                      class="infoSection MuiBox-root emotion-7"
                    >
                      <div
                        class="title MuiBox-root emotion-7"
                        data-se="authenticator-button-label"
                      >
                        IDP Authenticator
                      </div>
                      <div
                        class="description MuiBox-root emotion-7"
                        data-se="authenticator-button-description"
                      >
                        Redirect to verify with IDP Authenticator.
                      </div>
                      <div
                        class="cta-button actionName MuiBox-root emotion-7"
                        data-se="external_idp"
                      >
                        <span
                          class="MuiBox-root emotion-7"
                          data-se="cta-button-label"
                        >
                          Set up
                        </span>
                        <mocksvgicon />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="authButton MuiBox-root emotion-20"
                    data-se="authenticator-button"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiBox-root emotion-7"
                      data-se="authenticator-icon"
                    >
                      <div
                        class="iconContainer mfa-hotp MuiBox-root emotion-7"
                      >
                        <mocksvgicon />
                      </div>
                    </div>
                    <div
                      class="infoSection MuiBox-root emotion-7"
                    >
                      <div
                        class="title MuiBox-root emotion-7"
                        data-se="authenticator-button-label"
                      >
                        Atko Custom OTP Authenticator
                      </div>
                      <div
                        class="description MuiBox-root emotion-7"
                        data-se="authenticator-button-description"
                      >
                        Enter a temporary code generated from an authenticator device.
                      </div>
                      <div
                        class="cta-button actionName MuiBox-root emotion-7"
                        data-se="custom_otp"
                      >
                        <span
                          class="MuiBox-root emotion-7"
                          data-se="cta-button-label"
                        >
                          Set up
                        </span>
                        <mocksvgicon />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="authButton MuiBox-root emotion-20"
                    data-se="authenticator-button"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiBox-root emotion-7"
                      data-se="authenticator-icon"
                    >
                      <div
                        class="iconContainer mfa-symantec MuiBox-root emotion-7"
                      >
                        <mocksvgicon />
                      </div>
                    </div>
                    <div
                      class="infoSection MuiBox-root emotion-7"
                    >
                      <div
                        class="title MuiBox-root emotion-7"
                        data-se="authenticator-button-label"
                      >
                        Symantec VIP
                      </div>
                      <div
                        class="description MuiBox-root emotion-7"
                        data-se="authenticator-button-description"
                      >
                        Verify by entering a temporary code from the Symantec VIP app.
                      </div>
                      <div
                        class="cta-button actionName MuiBox-root emotion-7"
                        data-se="symantec_vip"
                      >
                        <span
                          class="MuiBox-root emotion-7"
                          data-se="cta-button-label"
                        >
                          Set up
                        </span>
                        <mocksvgicon />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="authButton MuiBox-root emotion-20"
                    data-se="authenticator-button"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiBox-root emotion-7"
                      data-se="authenticator-icon"
                    >
                      <div
                        class="iconContainer mfa-yubikey MuiBox-root emotion-7"
                      >
                        <mocksvgicon />
                      </div>
                    </div>
                    <div
                      class="infoSection MuiBox-root emotion-7"
                    >
                      <div
                        class="title MuiBox-root emotion-7"
                        data-se="authenticator-button-label"
                      >
                        YubiKey Authenticator
                      </div>
                      <div
                        class="description MuiBox-root emotion-7"
                        data-se="authenticator-button-description"
                      >
                        Verify your identity using YubiKey
                      </div>
                      <div
                        class="cta-button actionName MuiBox-root emotion-7"
                        data-se="yubikey_token"
                      >
                        <span
                          class="MuiBox-root emotion-7"
                          data-se="cta-button-label"
                        >
                          Set up
                        </span>
                        <mocksvgicon />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-10"
                >
                  <a
                    class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways emotion-137"
                    data-se="cancel"
                    href="javascript:void(0)"
                    inputref="[object Object]"
                  >
                    Back to sign in
                  </a>
                </div>
              </div>
            </form>
          </div>
          <div
            data-test="sentinelEnd"
            tabindex="0"
          />
        </div>
      </main>
    </span>
  </div>
</div>
`;
